File: 0383-ransom-note/0383-ransom-note.cpp
class Solution {
 public:
  bool canConstruct(string ransomNote, string magazine) {
    int map[26] = {0};
    for (int i = 0; i < magazine.length(); i++) {
      map[magazine[i] - 'a']++;
    }
    for (int i = 0; i < ransomNote.length(); i++) {
      int value = map[ransomNote[i] - 'a'];
      if (value <= 0) return false;
      map[ransomNote[i] - 'a']--;
    }
    return true;
  }
};

File: Python/0383-ransom-note.py
# time complexity: O(n)
# space complexity: O(1)
from collections import Counter

class Solution:
    def canConstruct(self, ransomNote: str, magazine: str) -> bool:
        ransomNoteCounter = Counter(ransomNote)
        magazineCounter = Counter(magazine)
        for key, freq in magazineCounter.items():
            if key in ransomNoteCounter:
                ransomNoteCounter[key] -= freq
                if ransomNoteCounter[key] <= 0:
                    del ransomNoteCounter[key]
        return len(ransomNoteCounter) == 0


ransomNote = "a"
magazine = "b"
print(Solution().canConstruct(ransomNote, magazine))
ransomNote = "aa"
magazine = "ab"
print(Solution().canConstruct(ransomNote, magazine))
ransomNote = "aa"
magazine = "aab"
print(Solution().canConstruct(ransomNote, magazine))
